### DISCUSSION TITLE
My first Lasercutter/engraver (patched LaserGRBL for Marlin M106/M107)

### DISCUSSION THREAD

**Discussion by TargetAlpha** [arkypita/LaserGRBL#1713] [Show and tell]
Hello there,

in the last couple of weeks i  have been busy building my first Lasercutter.

Since i was having a bit of trouble in some areas, i wanted to share what i did to get where i am now.

From the beginning:
Last Year, before Christmas, i decided to buy my first 3D Printer. An Elegoo Neptune 2.
And after a while the obvious question was: why not put a Laser on it?
So i ordered a 20W NEJE (N30820) Laser module.

But i didn't like the idea of attaching the Laser to the extruder. And i absolutely didn't like the idea of removing the extruder and attaching the Laser. I'm too lazy to change tools all the time.

So what i did instead... was buying a used Geeetech i3 pro B. I hacked it apart and rebuild it big enough to fit a piece of DIN A3 Paper.
For now i have done this with some scrap that was lying around. It is working but i plan to rebuild it a little nicer once all the kinks are worked out.

![IMG_20220109_125546688](https://user-images.githubusercontent.com/3617720/148686255-ce764038-03d0-4a54-8c5c-cf2291654e59.jpg)

The controller on this machine is a GT2560 , revision A+. It was running an older version of Marlin before and i since flashed Marlin 1.1.9.

I found the documentation on how to connect the Laser to the controller a bit confusing so i want to explain how i did it.
Here is some nice information from NEJE: https://wiki.nejetool.com/doku.php?id=laser_module_308020w

NEJE provides you with 2 different boards that you can use to connect power and the control signal to the Laser. I ended up using this one since it also allows me to control the Laser manually over the buttons. Which is very handy to adjust the focus.
![pwm_board](https://user-images.githubusercontent.com/3617720/148674762-d6318396-1f32-432b-8b71-50111758e6bb.jpg)
The Laser comes with it's own 12V powersupply and needs a PWM Signal to control the output power.
I think most people running these Lasers on 3D Printers use the PWM Signal of the Coolingfan.
You can easily control it with gcode and the wiring is already there.

So here is where it got a bit confusing for me.
NEJE labels the PWM Input as "TTL IN". To me, TTL means 5V logic. Since the Fan runs on 12V, i did not want to connect the PWM In directly. However, NEJE also specifies an input range from 3.3 to 12V though, so it should be fine..

Just one more problem. There is always +12V on the fanconnector. The fan is switched through a mosfet to ground.
So if the Laser is connected directly to the fan connector, that leaves us with 12V floating on the PWM In, even while the Fan is off.
This could propably lead to all kinds of funny behavior. I think i remember at least one person writing somewhere that his settings were inverted. At 0% his Laser was fully on, at 100% the Laser was off. This might have something to do with this.

So what i ended up doing was soldering the wires to the board to grab the 5V Signal directly from the controller.
That solved my worries about the voltage levels, gives me a proper Ground and a nicely switched 5V control signal.
I soldered to the resistor R15 since one side is connected to ground, the other to the Pin on the Atmega.
![Hardware_GT2560_RevA_2](https://user-images.githubusercontent.com/3617720/148675652-1c0cdf33-4f98-46bd-8849-cd66832ce7b6.png)
![IMG_20220109_125619800](https://user-images.githubusercontent.com/3617720/148685244-8e1cff7d-0a50-4a2b-a6f4-2837315c5cd6.jpg)

The next thing i needed was gcode.
A lot of people use inkscape together with the J Tech Photonics Laser Tool Plugin.
And it works great. You can specify your own Commands to turn the Laser off and on. And the code runs really well on my machine.
However, i really don't like working with inkscape. It just feels cumbersome. Opening the file, generate path, fiddle through the extensions menu, running the extension.. And i never had the energy to really figure out how to work with inkscape.

So then i found LaserGRBL. Using it feels way easier and more comfortable. 
But, as it's called LaserGRBL and not LaserMarlin, some things currently don't work as needed.

I will not discuss every change i made in detail since that would be a bit lengthy.
But i will attach my project folder and tell you the files i changed.
If there are specific questions, i will try to answer.

-GRBLFile.cs
-RasterConverter\ConvertSizeAndOptionForm.cs
-SvgConverter\ConvertSizeAndOptionForm.cs
-SvgConverter\gcoderelated.cs

[LaserGRBL-master.zip](https://github.com/arkypita/LaserGRBL/files/7835157/LaserGRBL-master.zip)

When you set the firmware-type to Marlin it is pretty much hardwired to use the gcodes M106 and M107.
When you change it back to GRBL ut should work with M3/4/5 again.
I did add M106 to the drop-down menus but i'm not skilled enough to fully integrate it.

There is also one thing where i want to ask  for help.
I tried to connect LaserGRBL to the machine over the usb serial interface.
My Marlin wants to communicate at 250.000baud. I think that is the default setting.
First: that option is not available in LaserGRBL. But i added it in ConnectLogForm.cs and it seems to communicate at the right speed.
LaserGRBL spams repeatedly M114 at the machine but does not get a response.
When i connect with putty to the same interface at 250.000baud it talks to me just fine and i get the expected answer to M114.

Can anyone confirm if this functions works for them in combination with Marlin?
I suspect that it might be a configuration problem with my Marlin setup.

Have a nice weekend and thanks for this great program.

**Comment by TargetAlpha** [comment]
Fixed it.
In GRBLCore.cs change

`						if (QueryTimer.Expired)
								QueryPosition();`

to 

`						if (QueryTimer.Expired)
							if (HasIncomingData())
							{
								QueryPosition();
							}`

Seems to work better with the Protocol Settings set to "USB Serial2".

**Reply by ivdor** [reply]
Hello. Could you upload somewhere a compiled version for marlin M106/107?
Or make pull request with it?

Sorry, I already find it in archive and will test it :)
WBR, Ivan

**Reply by ivdor** [reply]
@TargetAlpha 
Some problems in trace/passthrough modes.
Marlin bad works with "G0 X Y Sxxx" format in this gcodes
"Sxxx" must be in separate line "M106 Sxxx"

Trace/pass
> G90 (use absolute coordinates)
> G0 X0 Y0 F1000
> M106 S0
> G0 X1.588 Y0 S0
> G1 X9.79 S2
> G0 X17.462 Y0 S0
> G1 X19.579 S2
> X21.167 S3
> G0 X21.96 Y0 S0

Vector mode works good
> G90 (use absolute coordinates)
> M107
> G1 F1000
> G0 X2.924 Y23.6
> M106 S100
> G1 X1.762 Y23.6
> M106 S0
> G0 X1.748 Y23.8
> M106 S100
> G1 X2.923 Y23.8
> M106 S0

**Reply by TargetAlpha** [reply]
Hi there,

i see what you mean. I'll take a look if i can figure it out.
Until now i have only used vectorize so i didn't even notice.

**Reply by MadMark13** [reply]
I've got an old LaserBot with marlin (1.0.2) on it. 
Same connection issue with LaserGRBL 4.7.1 sending the M114 and not showing a response, where putty works fine.
My old version of marlin only seems to support M4 Pxxx codes to turn on the laser, but will turn off on only with M4 P0 or G0 commands.  Have to re-do the M4 to re-light the laser...

**Reply by TargetAlpha** [reply]
With my previous fix, the connect button was only working in debugging because the program was running much slower. I added a little delay and it's working for me now. I also changed the homing command from $H to G28.

Here is my current version. I'm afraid i can't really test whats working on older versions.
This is just what's working for me.

If you only want to fix the connect issue in the official version, take a look at core/grblcore.cs

You just need to change 

`if (QueryTimer.Expired)
						QueryPosition();`

to 			
		
`if (QueryTimer.Expired)
					{
						if (MachineStatus == MacStatus.Connecting && Tools.HiResTimer.TotalMilliseconds - connectStart > 1000)
							QueryPosition();
					}`

 at line 1569.

[LaserGRBL-master.zip](https://github.com/arkypita/LaserGRBL/files/7892784/LaserGRBL-master.zip)
.

**Reply by MadMark13** [reply]
I manged to get it talking with Marlin/Usb Serial/Synchronous/Quiet.
Thanks for the code.  G28 works for me.  I've got the buttons customized to use the code that works for me.
This older Marlin I have doesn't support the M106.  
It only seems to do M4 Pxxx  (0-255) where the diode stays lit through subsequent G1 moves, and turns off with a G0.
Needs another nonzero M4 Pxxx to re-enable after a G0.
I wonder if a table could be made to substitute M4 Pxxx for M3 Sxx and M5P0 for M5 like I did in the custom button. if Marlin version is <2.0 or therabouts.

**Comment by Johnnybeaner** [comment]
Hey Guys;
I downloaded and installed the latest revision (I think). I was able to update the optional buttons to turn my laser on and off. But the program is still using the GRBL  instead of the Marlin commands during the etching process.
Any idea what I'm doing wrong?

Thanks;
John

**Reply by arkypita** [reply]
Did you switch to Marlin? 
![image](https://user-images.githubusercontent.com/8782035/153365311-53210279-cace-4500-a6c7-939159ded279.png)

**Reply by Johnnybeaner** [reply]
Yes
Could I be using an older verson?

**Reply by Johnnybeaner** [reply]
I updated the software and the laser still won't come on.
The GCode uses M3, M4, M5 and not M106 & M107.

**Reply by ivdor** [reply]
Hello. Original program doesn't have m106/m107 support.
You can download archive from the first message, "lasergrbl-master" by Target Alpha.
This version can export gcode with M106 command support in vector mode

**Reply by Johnnybeaner** [reply]
Yes, I have but it's not compiled.
I've tried compiling it myself and it has errors.  Unfortunately, I'm not familiar with this language. 
Can someone put the compiled version on the site?
Thanks;
John

**Reply by ivdor** [reply]
First time I doesn't found it too)
Search compiled version in archive/lasergrbl/bin/debug/ folder

**Reply by Johnnybeaner** [reply]
Awesome!
Thanks for your help!!

**Reply by Johnnybeaner** [reply]
Hi Arkypita;
I tried this again without any luck.
In the attached picture you can see that I have the program set to Marlin, but the G-Code is using the GRBL (M3) commands.
Any suggestions?
Thanks;
John

![LaserGRBL Issue](https://user-images.githubusercontent.com/99362559/162590462-34e143f6-44e3-4d4e-8f17-9c70f8c5b457.png)
.

**Comment by CoefficientofSquish** [comment]
I'm running into the same issue at 115200, can connect with serial monitor in arduino IDE just fine, send commands and everything works as it should, but when I select Marlin in LaserGRBL and try to connect it just times out every time.  I will try to implement the solutions I saw here and see if that helps.  I went through my Marlin config and tried to find anything pertaining to the serial communication that could be causing it but so far nothing has helped.
